rfc/0010-constraint-expressions/README.md

```diff
@@ -0,0 +1,132 @@
+-->
+
+## Summary
+
+This RFC presents an alternative representation of value type constraints. It replaces the current solution.
+
+## Motivation
+
+Defining constraints is currently very bloated on the syntax level.
+
+Example:
+```
+constraint HundredScale oftype RangeConstraint {
+  lowerBound: 1;
+  lowerBoundInclusive: false;
+  upperBound: 100;      
+}
+```
+
+## Explanation
+
+Instead of using this block-like syntax, we can use expressions to define constraints.
+
+The general syntax looks like this: 
+```
+constraint <name> on <primitive value type> = <expression with parameter value, evaluating to boolean>
+```
+
+### Examples
+
+The following subsections will go over the existing constraints and showcase them in an example.
+
+#### AllowlistConstraint
+
+```
+constraint TimeUnitString on text = 
+  value =="ms" or 
+  value == "s" or 
+  value == "min" or 
+  value == "h" or 
+  value == "d" or 
+  value == "m" or 
+  value == "y";
+```
+
+#### DenylistConstraint
+```
+constraint NoPrimaryColors on text = 
+  not (value == "red" or value == "blue" or value == "yellow");
+
+// alternative
+constraint NoPrimaryColors on text = 
+  value != "red" and 
+  value != "blue" and 
+  value != "yellow";
+```
+
+#### LengthConstraint
+```
+constraint JavaStringLength on text = value.length  >= 0 and value.length  <= maxLength;
+```
+
+#### RangeConstraint
+```
+constraint HundredScale on decimal = value > 1 and value <= 100; 
+```
+
+#### RegexConstraint
+```
+constraint IFOPT_Format on text = value matches /[a-z]{2}:\d+:\d+(:\d+)?(:\d+)?/;
+```
+
+### Hidden Enhancements
+* Expressions allow parameters (keyword `value`).
+* The length of string parameters is accessible by `value.length`.
+* The assignment operator `=` instead of `:` because the latter looks off in that context.
+* The binary string operator `matches` is introduced.
+
+### Usage in ValueTypes
+In ValueTypes, these constraints can be added like before by name. Combining constraints on non-matching value types leads to an error at runtime in the IDE hints.
+```
+valuetype GasFillLevel oftype number {
+    constraints: [ GasFillLevelRange ];
+}
+```
+
+Alternatively, constraints can be defined inline with the caveat that they are not reusable elsewhere. Value then adheres to the type of the ValueType.
+```
+valuetype GasFillLevel oftype number {
+    constraints: [ value >= 0 and value <= 100 ];
+}
+```
+
+
+## Drawbacks
+
+- Inconsistency of constraint definition via `=` and property initialization via `:`.
+- Future constraints have to be expressible by a boolean expression.
+- Users now have to be able to write expressions instead of just supplying property values.
+
+## Alternatives
+
+- Syntax to define parameter in constraint could be different, e.g., `constraint MyConstraint = (number) => <expression>`.
+- Allow if/else statements in a more imperative way rather than packing everything into one expression.
```

Comment:
I think this could indeed be beneficial for cases where the expression becomes large and confusing. An alternative would be to nest constraints, so the overall complexity could be broken up into smaller sub-constraints.

As this is not an exclusive alternative, I'd shift this point to the `Possible Future Changes/Enhancements` section.

For now, I'd just go with the boolean expressions suggested in this RFC and observe whether the need for splitting up complex expressions arises at some point.